### DISCUSSION TITLE
rendering plot

### DISCUSSION THREAD

**Discussion by decareano** [theOehrly/Fast-F1#450] [General]
Hi, I am trying to get position changes for the japanese GP. Please see sample:

<img width="648" alt="Screen Shot 2023-09-24 at 10 06 35 AM" src="https://github.com/theOehrly/Fast-F1/assets/3332246/966302ca-a4f3-411e-8da3-b13c07b29b61">

when I run:

plt.show() nothing renders. Thanks.

**Comment by Casper-Guo** [comment]
Looks like the axes were not added to the current figure. What are your codes that came before the plotting calls?

**Comment by decareano** [comment]
Update: Better code read. Thanks


```
import fastf1.plotting
import matplotlib.pyplot as plt
fastf1.plotting.setup_mpl(misc_mpl_mods=False)

session = fastf1.get_session(2023, 16, 'R')
session.load(telemetry=False, weather=False)

fig, ax = plt.subplots(figsize=(8.0, 4.9))

for drv in session.drivers:
    drv_laps = session.laps.pick_driver(drv)

    abb = drv_laps['Driver'].iloc[0]
    color = fastf1.plotting.driver_color(abb)

    ax.plot(drv_laps['LapNumber'], drv_laps['Position'],
            label=abb, color=color)

ax.set_ylim([20.5, 0.5])
ax.set_yticks([1, 5, 10, 15, 20])
ax.set_xlabel('Lap')
ax.set_ylabel('Position')

ax.legend(bbox_to_anchor=(1.0, 1.02))
plt.tight_layout()

plt.show()

```

**Reply by theOehrly** [reply]
Does this code work now? It works for me. I haven't tried it in Jupyter, though.

**Comment by decareano** [comment]
Ahhh. It works in Visual Studio. I usually work with Jupyter. This is what I noticed in Jupyter. I will run a sample from the website (FastF1). First time running ok. On the second run and without making ANY changes, the sample will not run or it will run with errors in the rendering. I noticed it in about three samples by now. Any ideas?

**Reply by theOehrly** [reply]
I basically never use Jupyter. Do you rerun all the cells starting from ``fig, ax = plt.subplots(figsize=(8.0, 4.9))``?
IMO, you can't "show" a figure twice with Matplotlib. So you need to fully recreate it. Maybe that's the problem?

**Comment by decareano** [comment]
Yes, I usually run ALL the cells just to be sure. No issues. I can use either one. 
Now, I want to be able to use Plotly for better rendering. I understand I can "convert" a Matplotlib into a Plotly graph.
Please see below: I added two lines (they are commented out with: "// ADDED". However, I don't see any difference when it renders. 
Thanks,
```
import fastf1 as ff1
from fastf1 import plotting
from fastf1 import utils
import chart_studio.plotly as py            // ADDED 
from matplotlib import pyplot as plt
from matplotlib.pyplot import figure
plt.style.use('dark_background')

import numpy as np
import pandas as pd

year, grand_prix, session = 2023, 'Japan', 'Q'

quali = ff1.get_session(year, grand_prix, session)
quali.load() # This is new with Fastf1 v.2.2


driver_1, driver_2 = 'VER', 'PIA'

laps_driver_1 = quali.laps.pick_driver(driver_1)
laps_driver_2 = quali.laps.pick_driver(driver_2)


fastest_driver_1 = laps_driver_1.pick_fastest()
fastest_driver_2 = laps_driver_2.pick_fastest()
print(fastest_driver_1)
print(fastest_driver_2)


# Retrieve the telemetry and add the distance column
telemetry_driver_1 = fastest_driver_1.get_telemetry().add_distance()
telemetry_driver_2 = fastest_driver_2.get_telemetry().add_distance()

team_driver_1 = fastest_driver_1['Team']
team_driver_2 = fastest_driver_2['Team']

delta_time, ref_tel, compare_tel = utils.delta_time(fastest_driver_1, fastest_driver_2)

plot_size = [50, 50]
mpl_fig = plt.figure()
plot_title = f"{quali.event.year} {quali.event.EventName} - {quali.name} - {driver_1} VS {driver_2}"
plot_ratios = [1, 3, 2, 1, 1, 2, 1]
plot_filename = plot_title.replace(" ", "") + ".png"

# Make plot a bit bigger
plt.rcParams['figure.figsize'] = plot_size

# Create subplots with different sizes
fig, ax = plt.subplots(7, gridspec_kw={'height_ratios': plot_ratios})

# Set the plot title
ax[0].title.set_text(plot_title)


# Delta line
ax[0].plot(ref_tel['Distance'], delta_time)
ax[0].axhline(0)
ax[0].set(ylabel=f"Gap to {driver_2} (s)")

# Speed trace
ax[1].plot(telemetry_driver_1['Distance'], telemetry_driver_1['Speed'] ,label=driver_1, color=ff1.plotting.team_color(team_driver_1))
ax[1].plot(telemetry_driver_2['Distance'], telemetry_driver_2['Speed'], label=driver_2, color=ff1.plotting.team_color(team_driver_2))
ax[1].set(ylabel='Speed')


ax[1].legend(loc="lower right")

# Throttle trace
ax[2].plot(telemetry_driver_1['Distance'], telemetry_driver_1['Throttle'], label=driver_1, color=ff1.plotting.team_color(team_driver_1))
ax[2].plot(telemetry_driver_2['Distance'], telemetry_driver_2['Throttle'], label=driver_2, color=ff1.plotting.team_color(team_driver_2))
ax[2].set(ylabel='Throttle')

# Brake trace
ax[3].plot(telemetry_driver_1['Distance'], telemetry_driver_1['Brake'], label=driver_1, color=ff1.plotting.team_color(team_driver_1))
ax[3].plot(telemetry_driver_2['Distance'], telemetry_driver_2['Brake'], label=driver_2, color=ff1.plotting.team_color(team_driver_2))
ax[3].set(ylabel='Brake')

# Gear trace
ax[4].plot(telemetry_driver_1['Distance'], telemetry_driver_1['nGear'], label=driver_1, color=ff1.plotting.team_color(team_driver_1))
ax[4].plot(telemetry_driver_2['Distance'], telemetry_driver_2['nGear'], label=driver_2, color=ff1.plotting.team_color(team_driver_2))
ax[4].set(ylabel='Gear')

# RPM trace
ax[5].plot(telemetry_driver_1['Distance'], telemetry_driver_1['RPM'], label=driver_1, color=ff1.plotting.team_color(team_driver_1))
ax[5].plot(telemetry_driver_2['Distance'], telemetry_driver_2['RPM'], label=driver_2, color=ff1.plotting.team_color(team_driver_2))
ax[5].set(ylabel='RPM')

# DRS trace
ax[6].plot(telemetry_driver_1['Distance'], telemetry_driver_1['DRS'], label=driver_1, color=ff1.plotting.team_color(team_driver_1))
ax[6].plot(telemetry_driver_2['Distance'], telemetry_driver_2['DRS'], label=driver_2, color=ff1.plotting.team_color(team_driver_2))
ax[6].set(ylabel='DRS')
ax[6].set(xlabel='Lap distance (meters)')


# Hide x labels and tick labels for top plots and y ticks for right plots.
for a in ax.flat:
    a.label_outer()

# Store figure
plt.savefig(plot_filename, dpi=400)
plt.show()

unique_url = py.plot_mpl(mpl_fig, filename="sin(x) test plot")      //ADDED
```

**Reply by Ivan-Keli** [reply]
I ran this code and got something totally unexpected

**Reply by Ivan-Keli** [reply]
Let me copy paste the code here then i'll screen record the output the results and the error before sending them in a zip file since it will exceed the 10MB limit.

1. Code

import fastf1 as ff1
from fastf1 import plotting
from fastf1 import utils
import chart_studio.plotly as py            #// ADDED 
from matplotlib import pyplot as plt
from matplotlib.pyplot import figure
plt.style.use('dark_background')

import numpy as np
import pandas as pd

year, grand_prix, session = 2023, 'Japan', 'Q'

quali = ff1.get_session(year, grand_prix, session)
quali.load() # This is new with Fastf1 v.2.2


driver_1, driver_2 = 'VER', 'PIA'

laps_driver_1 = quali.laps.pick_driver(driver_1)
laps_driver_2 = quali.laps.pick_driver(driver_2)


fastest_driver_1 = laps_driver_1.pick_fastest()
fastest_driver_2 = laps_driver_2.pick_fastest()
print(fastest_driver_1)
print(fastest_driver_2)


# Retrieve the telemetry and add the distance column
telemetry_driver_1 = fastest_driver_1.get_telemetry().add_distance()
telemetry_driver_2 = fastest_driver_2.get_telemetry().add_distance()

team_driver_1 = fastest_driver_1['Team']
team_driver_2 = fastest_driver_2['Team']

delta_time, ref_tel, compare_tel = utils.delta_time(fastest_driver_1, fastest_driver_2)

plot_size = [50, 50]
mpl_fig = plt.figure()
plot_title = f"{quali.event.year} {quali.event.EventName} - {quali.name} - {driver_1} VS {driver_2}"
plot_ratios = [1, 3, 2, 1, 1, 2, 1]
plot_filename = plot_title.replace(" ", "") + ".png"

# Make plot a bit bigger
plt.rcParams['figure.figsize'] = plot_size

# Create subplots with different sizes
fig, ax = plt.subplots(7, gridspec_kw={'height_ratios': plot_ratios})

# Set the plot title
ax[0].title.set_text(plot_title)


# Delta line
ax[0].plot(ref_tel['Distance'], delta_time)
ax[0].axhline(0)
ax[0].set(ylabel=f"Gap to {driver_2} (s)")

# Speed trace
ax[1].plot(telemetry_driver_1['Distance'], telemetry_driver_1['Speed'] ,label=driver_1, color=ff1.plotting.team_color(team_driver_1))
ax[1].plot(telemetry_driver_2['Distance'], telemetry_driver_2['Speed'], label=driver_2, color=ff1.plotting.team_color(team_driver_2))
ax[1].set(ylabel='Speed')


ax[1].legend(loc="lower right")

# Throttle trace
ax[2].plot(telemetry_driver_1['Distance'], telemetry_driver_1['Throttle'], label=driver_1, color=ff1.plotting.team_color(team_driver_1))
ax[2].plot(telemetry_driver_2['Distance'], telemetry_driver_2['Throttle'], label=driver_2, color=ff1.plotting.team_color(team_driver_2))
ax[2].set(ylabel='Throttle')

# Brake trace
ax[3].plot(telemetry_driver_1['Distance'], telemetry_driver_1['Brake'], label=driver_1, color=ff1.plotting.team_color(team_driver_1))
ax[3].plot(telemetry_driver_2['Distance'], telemetry_driver_2['Brake'], label=driver_2, color=ff1.plotting.team_color(team_driver_2))
ax[3].set(ylabel='Brake')

# Gear trace
ax[4].plot(telemetry_driver_1['Distance'], telemetry_driver_1['nGear'], label=driver_1, color=ff1.plotting.team_color(team_driver_1))
ax[4].plot(telemetry_driver_2['Distance'], telemetry_driver_2['nGear'], label=driver_2, color=ff1.plotting.team_color(team_driver_2))
ax[4].set(ylabel='Gear')

# RPM trace
ax[5].plot(telemetry_driver_1['Distance'], telemetry_driver_1['RPM'], label=driver_1, color=ff1.plotting.team_color(team_driver_1))
ax[5].plot(telemetry_driver_2['Distance'], telemetry_driver_2['RPM'], label=driver_2, color=ff1.plotting.team_color(team_driver_2))
ax[5].set(ylabel='RPM')

# DRS trace
ax[6].plot(telemetry_driver_1['Distance'], telemetry_driver_1['DRS'], label=driver_1, color=ff1.plotting.team_color(team_driver_1))
ax[6].plot(telemetry_driver_2['Distance'], telemetry_driver_2['DRS'], label=driver_2, color=ff1.plotting.team_color(team_driver_2))
ax[6].set(ylabel='DRS')
ax[6].set(xlabel='Lap distance (meters)')


# Hide x labels and tick labels for top plots and y ticks for right plots.
for a in ax.flat:
    a.label_outer()

# Store figure
plt.savefig(plot_filename, dpi=400)
plt.show()

unique_url = py.plot_mpl(mpl_fig, filename="sin(x) test plot")      #//ADDED

2. Result
3. Error
[rendering plot - Copy.zip](https://github.com/theOehrly/Fast-F1/files/14234253/rendering.plot.-.Copy.zip)

**Reply by theOehrly** [reply]
Nothing there looks incorrect, I'd say. The additional text output is because of the``print(...)`` calls. The plots are darker because a different matplotlib theme is configured, and the font looks smaller, because the plot size is configured to be very large. Everything looks correct.

**Reply by Ivan-Keli** [reply]
Thank you very much 🙏🏽

**Comment by theOehrly** [comment]
I tried to reproduce this just now, but I can't. Could you maybe create a notebook on Google Colab or something like that, that you can easily share, so I can look at this. 
But to be honest, I don't think this has anything to do with FastF1. It's probably something about Matplotlib and the (version of?) Jupyter.

**Comment by decareano** [comment]
FYI. Let me know if you could open it.
https://github.com/decareano/F1Repo/blob/main/japanQVerPia.ipynb

**Reply by theOehrly** [reply]
Yes, I can access it. I tried running the notebook on Google Colab. It worked exactly like I expected it to work.  I could rerun cells and plotting still worked. 
I'm still not sure, what the problem is that you are having. Are you using Jupyter locally? If so, please try to reproduce your issue on a platform like Google Colab to see whether it is related to your environment or not.
